src/object/srv_obj.c

```diff
@@ -1898,6 +1898,65 @@ ds_obj_tgt_update_handler(crt_rpc_t *rpc)
 	obj_ioc_end(&ioc, rc);
 }
 
+static int
+obj_dth_agg_cb(struct dtx_leader_handle *dlh, void *agg_arg)
+{
+	uint64_t	*flag = agg_arg;
+	int		succeeds = 0;
+	int		allow_failure = 0;
+	int		allow_failure_cnt = 0;
+	int		result = 0;
+	int		i;
+
+	D_ASSERT(flag != NULL);
+	if (*flag && (DAOS_COND_PUNCH | DAOS_COND_DKEY_UPDATE |
+		     DAOS_COND_AKEY_UPDATE))
+		allow_failure = -DER_NONEXIST;
+	else if (*flag && (DAOS_COND_AKEY_INSERT | DAOS_COND_DKEY_INSERT))
+		allow_failure = -DER_EXIST;
+
+	for (i = 0; i < dlh->dlh_sub_cnt; i++) {
+		struct dtx_sub_status	*sub = &dlh->dlh_subs[i];
+
+		if (sub->dss_result == 0) {
+			succeeds++;
+		} else if (sub->dss_result == allow_failure) {
+			allow_failure_cnt++;
+		} else {
+			if (result == -DER_INPROGRESS || result == 0)
+				result = sub->dss_result;
+		}
+	}
+
+	D_DEBUG(DB_IO, DF_DTI" %d/%d shards flags "DF_X64" result %d\n",
+		DP_DTI(&dlh->dlh_handle.dth_xid), allow_failure_cnt,
+		succeeds, *flag, result);
+
+	if (*flag && (DAOS_COND_PUNCH | DAOS_COND_DKEY_UPDATE |
+		      DAOS_COND_AKEY_UPDATE)) {
+		/* For punch/update, if there is one shard succeeds, then
+		 * let's ignore DER_NONEXIST, since the object may not
+		 * exist on some shards due to EC partial udpate.
+		 */
+		if (result == 0 && succeeds == 0)
+			return -DER_NONEXIST;
```

Comment:
the problem is this doesn't work.   We don't want -DER_NONEXIST in this case, but rather we want to do the update on those shards